### ISSUE TITLE
batch_end_callback doesn't show eval_metric

### ISSUE THREAD

**Issue by zihaolucky** [dmlc/mxnet#3062]
I want to implement CNN for text on my own but failed to get evaluation metric in batch_end_callback like this(not exactly top-k, its just accuracy):

```
2016-08-12 20:15:43,819 Node[0] Epoch[5] Batch [50]    Speed: 444.47 samples/sec    Train-accuracy=0.844219
2016-08-12 20:15:43,819 Node[0] Epoch[5] Batch [50]    Speed: 444.47 samples/sec    Train-top_k_accuracy_5=0.994219
2016-08-12 20:15:43,820 Node[0] Epoch[5] Batch [50]    Speed: 444.47 samples/sec    Train-top_k_accuracy_10=1.000000
2016-08-12 20:15:43,820 Node[0] Epoch[5] Batch [50]    Speed: 444.47 samples/sec    Train-top_k_accuracy_20=1.000000
2016-08-12 20:15:58,518 Node[0] Epoch[5] Batch [100]    Speed: 435.42 samples/sec    Train-accuracy=0.846094
2016-08-12 20:15:58,519 Node[0] Epoch[5] Batch [100]    Speed: 435.42 samples/sec    Train-top_k_accuracy_5=0.995625
2016-08-12 20:15:58,519 Node[0] Epoch[5] Batch [100]    Speed: 435.42 samples/sec    Train-top_k_accuracy_10=1.000000
2016-08-12 20:15:58,519 Node[0] Epoch[5] Batch [100]    Speed: 435.42 samples/sec    Train-top_k_accuracy_20=1.000000
2016-08-12 20:16:13,001 Node[0] Epoch[5] Batch [150]    Speed: 441.95 samples/sec    Train-accuracy=0.860938
```

Here's my network:

```python
10 def get_net(vocab_size, embedding_size, sentence_size, batch_size, num_classes, with_embedding, dropout=0.0, filter_list=[3,4,5], num_filter=100):
 11     """
 12     Network structure.
 13     """
 14     import logging
 15     head = head = '%(asctime)-15s %(message)s'
 16     logging.basicConfig(level=logging.DEBUG, format=head)
 17
 18     data = mx.symbol.Variable('data')
 19     label = mx.symbol.Variable('softmax_label')
 20
 21     # embedding layer
 22     if not with_embedding:
 23         word_embed = mx.symbol.Embedding(data=data, input_dim=vocab_size,
 24                                       output_dim=embedding_size, name='word_embedding')
 25         conv_input = mx.symbol.Reshape(data=word_embed, target_shape=(batch_size, 1, sentence_size, embedding_size))  # convolution layer needs 4D input.
 26     else:
 27         logging.info('with pretrained embedding.')
 28         conv_input = data
 29
 30     # convolution and pooling layer
 31     pooled_outputs = []
 32     for i, filter_size in enumerate(filter_list):
 33         convi = mx.symbol.Convolution(data=conv_input, kernel=(filter_size, embedding_size),
 34                                    num_filter=num_filter)
 35         acti  = mx.symbol.Activation(data=convi, act_type='relu')
 36         pooli = mx.symbol.Pooling(data=acti, pool_type='max', kernel=(sentence_size - filter_size + 1, 1), stride=(1,1))  # max pooling on entire sentence feature ma    p.
 37         pooled_outputs.append(pooli)
 38
 39     # combine all pooled outputs
 40     num_feature_maps = num_filter * len(filter_list)
 41     concat = mx.symbol.Concat(*pooled_outputs, dim=1)  # max-overtime pooling. concat all feature maps into a long feature before feeding into final dropout and full    y connected layer.
 42     h_pool = mx.symbol.Reshape(data=concat, shape=(batch_size, num_feature_maps))  # make it flat/horizontal
 43
 44     # dropout
 45     if dropout > 0.0:
 46         logging.info('use dropout.')
 47         drop = mx.symbol.Dropout(data=h_pool, p=dropout)
 48     else:
 49         logging.info('Do not use dropout.')
 50         drop = h_pool
 51
 52     # fully connected and softmax output.
 53     fc = mx.symbol.FullyConnected(data=drop, num_hidden=num_classes)
 54     softmax = mx.symbol.SoftmaxOutput(data=fc, label=label)
 55
 56     return softmax
 57

```

I use DataIter in which the shape in `provide_data` and `provide_label` are same as the netowrk's :

```
2016-08-18 22:24:20,943 provide_data. [('data', (64L, 1L, 19L, 128L))]
2016-08-18 22:24:20,943 provide_label. [('softmax_label', (64L,))]
```

```python
 66 class DataIter(mx.io.DataIter):
 67     """
 68     Define data loading and batch-iter related functionality.
 69     """
 70     def __init__(self, batch_size, num_classes, embedding_size, num_epochs,
 71                  input_file, word2vec_file, with_embedding):
 72         super(DataIter, self).__init__()
...
...
 92         if with_embedding:
 93             # word2vec
 94             self.word2vec = data_helpers.load_pretrained_word2vec(word2vec_file)
 95             self.vocab_size = len(self.word2vec)
 96             # text image
 97             self.data, self.label = data_helpers.load_data_with_word2vec(input_file, self.word2vec)
 98             self.sentence_size = self.data.shape[1]
 99             self.provide_data = [('data', (self.batch_size, 1, self.sentence_size, self.embedding_size))]
100             self.provide_label = [('softmax_label', (self.batch_size, ))]
101         else:
102             # training sample
103             self.data, self.label, self.vocab, self.vocab_inv = data_helpers.load_data(input_file)
104             self.vocab_size = len(self.vocab)
105             self.sentence_size = self.data.shape[1]
106             self.provide_data = [('data', (self.batch_size, self.sentence_size))]
107             self.provide_label = [('softmax_label', (self.batch_size, ))]
```

```python
186     data_train = DataIter(batch_size, num_classes, embedding_size, num_epoch,
187                           input_file, word2vec_file, with_embedding=True)
190
191
192     # get text_cnn network and model
193     cnn = get_net(vocab_size=data_train.vocab_size, embedding_size=embedding_size,
194                   sentence_size=data_train.sentence_size, batch_size=batch_size,
195                   num_classes=num_classes, with_embedding=True, dropout=0.5)
196
197
198     model = mx.model.FeedForward(ctx = mx.gpu(0),
199                                  symbol = cnn,
200                                  num_epoch = num_epoch,
201                                  learning_rate = 0.3,
202                                  momentum = 0.9,
203                                  wd = 0.0000,
204                                  initializer = mx.init.Xavier(factor_type='in', magnitude=2.34))
205
206     eval_metrics = ["acc"]
209
210     # training
211     model.fit(X                  = data_train,
212               #eval_data                 = data_eval,
213               eval_metric        = eval_metrics,
214               batch_end_callback = mx.callback.Speedometer(batch_size, 50))
```

It makes me confused why evaluation metrics doesn't show up. I suppose it should have even though I do not provide `eval_data`.

**Comment by zihaolucky** [comment]
After I post this issue, the report came out surprisingly. Close the issue.
